### PR TITLE
Print messages when error happens

### PR DESCRIPTION
**Pull request by rtakacs**
Print error messages to be more informative.

### CODE REVIEW

File: jerry-core/ecma/operations/ecma-exceptions.cpp [hunk 1]
```diff
@@ -29,6 +31,76 @@
  * @{
  */
 
+void
+ecma_print_error_messages (ecma_object_t *obj_p)
```
Comment (ruben-ayrapetyan) [inline jerry-core/ecma/operations/ecma-exceptions.cpp]: Please, add comments.

Comment (ruben-ayrapetyan) [inline jerry-core/ecma/operations/ecma-exceptions.cpp]: The function is better to be named like `ecma_print_exception_message`

**Comment by ruben-ayrapetyan** [comment]
@rtakacs, could you, please, also implement pass of SyntaxError's / ReferenceError's messages from parser to corresponding exception object, and turn off print of the message during parser operation?

See also:
 #724, #726

**Comment by rtakacs** [comment]
@ruben-ayrapetyan: of course :)

**Comment by LaszloLango** [comment]
@rtakacs, please update to the current master.

**Comment by rtakacs** [comment]
@LaszloLango: Before I update this PR I would like to add error object logic to the parser as well.

**Comment by rtakacs** [comment]
I've uploaded a WIP of my work. Any suggestions are welcome. It works for file and for REPL as well.

Some output: 

```
> a = 1;
> b ====
SyntaxError: Primary expression expected. [Line: 2, Column: 6]
```
```
> foo
ReferenceError: foo is not defined.
```

**Comment by LaszloLango** [comment]
@rtakacs, I think the string concatenation should be on the client side. The core should return with an error object like:
```C++
struct error_obj {
  uint8_t type;
  ecma_string_t *msg;
  uint16_t line;
  uint16_t column;
};
```

Similar to `parser_error_location` used in `parser_parse_source`

File: jerry-core/jerry.c [hunk 2]
```diff
@@ -2055,10 +2058,17 @@ jerry_parse_and_save_snapshot (const jerry_api_char_t *source_p, /**< script sou
...
 
   if (is_for_global)
   {
-    parse_status = parser_parse_script (source_p, source_size, &bytecode_data_p);
+    parse_status = parser_parse_script (source_p, source_size, &bytecode_data_p, &error_obj_p);
+
+    if (parse_status != JSP_STATUS_OK)
+    {
+      JERRY_ASSERT (error_obj_p != NULL);
+      ecma_deref_object (error_obj_p);
```
Comment (zherczeg) [inline jerry-core/jerry.c]: Deref object?

File: jerry-core/parser/js/js-parser.c [hunk 3]
```diff
@@ -2254,19 +2254,22 @@ parser_set_show_instrs (int show_instrs) /**< flag indicating whether to dump by
...
-
 /**
  * Parse EcamScript source code
  */
 jsp_status_t
 parser_parse_script (const jerry_api_char_t *source_p, /**< source code */
                      size_t size, /**< size of the source code */
-                     ecma_compiled_code_t **bytecode_data_p) /**< result */
+                     ecma_compiled_code_t **bytecode_data_p, /**< [out] JS bytecode */
+                     jerry_api_object_t **error_obj_p __attr_unused___) /**< [out] error object */
```
Comment (zherczeg) [inline jerry-core/parser/js/js-parser.c]: Why unused?

File: tests/unit/test-api.c [hunk 4]
```diff
@@ -332,13 +332,14 @@ main (void)
   jerry_api_object_t *global_obj_p, *obj_p;
   jerry_api_object_t *external_func_p, *external_construct_p;
   jerry_api_object_t *throw_test_handler_p;
+  jerry_api_object_t *err_obj_p = NULL;
   jerry_api_value_t res, args[2];
   char buffer[32];
 
-  is_ok = jerry_parse ((jerry_api_char_t *) test_source, strlen (test_source));
+  is_ok = jerry_parse ((jerry_api_char_t *) test_source, strlen (test_source), &err_obj_p);
   JERRY_ASSERT (is_ok);
```
Comment (zherczeg) [inline tests/unit/test-api.c]: is_ok && err_obj_p == NULL

File: jerry-core/jerry.c (hunk 2 above)
Comment (LaszloLango) [inline jerry-core/jerry.c]: If an unhandled exception occurs, then the error object parameter will be non-null and someone must to free it.

File: tests/unit/test-api.c [hunk 5]
```diff
@@ -332,13 +332,14 @@ main (void)
...
 
-  is_ok = (jerry_run () == JERRY_COMPLETION_CODE_OK);
+  is_ok = (jerry_run (&err_obj_p) == JERRY_COMPLETION_CODE_OK);
   JERRY_ASSERT (is_ok);
```
Comment (zherczeg) [inline tests/unit/test-api.c]: Ditto. Please check that the err_obj_p is NULL.

File: jerry-core/parser/js/js-parser.c (hunk 3 above)
Comment (LaszloLango) [inline jerry-core/parser/js/js-parser.c]: Compiler complains about it is set but not used.

**Comment by LaszloLango** [comment]
@zherczeg, updated

File: jerry-core/parser/js/js-parser.c (hunk 3 above)
Comment (zherczeg) [inline jerry-core/parser/js/js-parser.c]: Where? It seems to me a bug if we create an object and don't use (free) it.

Comment (LaszloLango) [inline jerry-core/parser/js/js-parser.c]: This is an output parameter. What do you suggest? If we free it here, then we are not able to print it in `main-unix`

File: tests/unit/test-api.c [hunk 6]
```diff
@@ -332,14 +332,15 @@ main (void)
...
   jerry_api_object_t *external_func_p, *external_construct_p;
   jerry_api_object_t *throw_test_handler_p;
+  jerry_api_object_t *err_obj_p = NULL;
   jerry_api_value_t res, args[2];
   char buffer[32];
 
-  is_ok = jerry_parse ((jerry_api_char_t *) test_source, strlen (test_source));
-  JERRY_ASSERT (is_ok);
+  is_ok = jerry_parse ((jerry_api_char_t *) test_source, strlen (test_source), &err_obj_p);
+  JERRY_ASSERT (is_ok && err_obj_p == NULL);
```
Comment (akiss77) [inline tests/unit/test-api.c]: Note: It's better practice to split and'ed asserts into two. In release, two asserts cost the same as one (i.e., nothing). However, if it comes to debugging, it's easier if the assert points exactly at source of failure. I.e., with the `&&` syntax, there will be a need for an extra round of execution to find out whether the `is_ok` or the `err_obj_p == NULL` part failed. With two asserts, that's not an issue.

**Comment by LaszloLango** [comment]
@zherczeg, updated

File: jerry-core/jerry.c [hunk 7]
```diff
@@ -2361,7 +2372,14 @@ jerry_exec_snapshot (const void *snapshot_p, /**< snapshot */
...
 
-    ret_code = vm_run_global ();
+    ecma_object_t *error_obj_p = NULL;
+
+    ret_code = vm_run_global (&error_obj_p);
+
+    if (ret_code == JERRY_COMPLETION_CODE_UNHANDLED_EXCEPTION)
+    {
+      ecma_deref_object (error_obj_p);
+    }
```
Comment (zherczeg) [inline jerry-core/jerry.c]: There should be an ASSERT here that error_obj_p is NULL or not-NULL depending on the ret_code

File: tests/unit/test-api.c (hunk 6 above)
Comment (zherczeg) [inline tests/unit/test-api.c]: I don't mind.

**Comment by zherczeg** [comment]
LGTM after the issues above fixed.

**Pull request merged**